### DISCUSSION TITLE
UOR Event Engine SDK Planning

### DISCUSSION THREAD

**Discussion by afflom** [emporous/community#3] [General]
UOR Event Engines are the imported application logic of a UOR Collection. An SDK may be useful in assisting adopters with writing UOR integrations for their existing applications. I have a few questions to start this conversation:

1. Is an SDK needed?
2. What will the SDK provide to adopters/integrators?

My thought of what the SDK could be is an application that can:
1. Be called in CI like a GH action
2. Enumerate the command tree of the importing code
3. Create an attribute to UOR events API mapping
4. Add the UOR API and mapping to the importing code
5. Compile and publish the modified application as UOR event engine
6. Using command tree enumeration and from step 2, the event engine publishing location from step 5, and a user defined config file; create a UOR schema and publish it.

This is a different workflow than a UOR native application would use. UOR native applications would be developed with imported UOR Client libraries.

**Comment by jpower432** [comment]
I think an SDK could be useful over something like a configuration file or Containerfile-like solution if we can show that the Event Engines would require an established pattern (similar to the Operator pattern) and required library usage.

**Comment by afflom** [comment]
Thinking more about this, I think that we could put a proxy between the UOR client and the Event Engine. The proxy would translate attributes to the Event Engine's command tree. The proxy would be generic to UOR and dynamically configured by the attribute mapping in the schema. 

This could be used as a compatibility layer for non-UOR native applications. UOR native applications would implement the UOR Framework and would not need a proxy. 

wdyt?

**Reply by usrbinkat** [reply]
This application of a proxy begins to sound like a sidecar proxy when looking at it in both a distributed system and in multiple independent nodes.

I agree a proxy could contain a mutation API to provide backward compatibility between UOR and legacy applications.

Still, I believe there is more utility in a UOR proxy, whether in front of a local cache or in front of a centralized UOR registry. If written as a few WASM plugins run via envoy with local-cache read/write access, we could instantiate a UOR native data mesh, add local cache TTL features to improve object network efficiency and facilitate the mutation webhook to mediate between legacy applications and uor objects.

The proxy discussion likely deserves a separate discussions thread as it is only loosely related to the SDK discussion.

**Reply by jpower432** [reply]
Great thoughts here. Though when I think of a proxy between the client and an event engine, my brain goes to an admission controller. I continued this on #5.
